File: weyl-core/src/main/java/io/quartic/weyl/core/compute/SpatialPredicateComputation.java
package io.quartic.weyl.core.compute;

import com.google.common.collect.ImmutableList;
import io.quartic.common.SweetStyle;
import io.quartic.weyl.core.model.Layer;
import io.quartic.weyl.core.model.LayerId;
import io.quartic.weyl.core.model.LayerMetadata;
import io.quartic.weyl.core.model.LayerPopulator;
import io.quartic.weyl.core.model.LayerSpec;
import io.quartic.weyl.core.model.LayerUpdate;
import io.quartic.weyl.core.model.NakedFeature;
import org.immutables.value.Value;
import rx.Observable;

import java.time.Clock;
import java.util.Collection;
import java.util.List;

import static io.quartic.weyl.api.LayerUpdateType.REPLACE;
import static io.quartic.weyl.core.live.LayerView.IDENTITY_VIEW;
import static java.util.stream.Collectors.toList;

@SweetStyle
@Value.Immutable
public abstract class SpatialPredicateComputation implements LayerPopulator {
    protected abstract LayerId layerId();
    protected abstract SpatialPredicateSpec spatialPredicateSpec();

    @Value.Default
    protected Clock clock() {
        return Clock.systemUTC();
    }

    @Override
    public List<LayerId> dependencies() {
        return ImmutableList.of(spatialPredicateSpec().getLayerA(), spatialPredicateSpec().getLayerB());
    }

    @Override
    public LayerSpec spec(List<Layer> dependencies) {
        final Layer layerA = dependencies.get(0);
        final Layer layerB = dependencies.get(1);

        return new LayerSpec(
                layerId(),
                new LayerMetadata(
                        name(layerA, layerB, spatialPredicateSpec()),
                        name(layerA, layerB, spatialPredicateSpec()),
                        layerA.getSpec().getMetadata().getAttribution(),
                        clock().instant()
                ),
                IDENTITY_VIEW,
                layerA.getSpec().getStaticSchema(),
                true
        );
    }

    private static String name(Layer layerA, Layer layerB, SpatialPredicateSpec spatialPredicateSpec) {
        return String.format("%s %s %s",
                layerA.getSpec().getMetadata().getName(),
                spatialPredicateSpec.getPredicate(),
                layerB.getSpec().getMetadata().getName());
    }

    @Override
    public Observable<LayerUpdate> updates(List<Layer> dependencies) {
        final Layer layerA = dependencies.get(0);
        final Layer layerB = dependencies.get(1);

        Collection<NakedFeature> bufferedFeatures = layerA.getIndexedFeatures().parallelStream()
                .filter(featureA -> layerB.getFeatures().stream()
                        .anyMatch(featureB -> spatialPredicateSpec().getPredicate()
                                        .test(featureA.getPreparedGeometry(), featureB.getGeometry())))
                .map(feature -> new NakedFeature(
                        feature.getFeature().getEntityId().getUid(),
                        feature.getFeature().getGeometry(),
                        feature.getFeature().getAttributes())
                )
                .collect(toList());

        return Observable.<LayerUpdate>never().startWith(new LayerUpdate(REPLACE, bufferedFeatures));
    }
}

File: weyl-core/src/main/java/io/quartic/weyl/core/compute/SpatialPredicateComputation.kt
package io.quartic.weyl.core.compute

import io.quartic.weyl.api.LayerUpdateType.REPLACE
import io.quartic.weyl.core.live.LayerView.IDENTITY_VIEW
import io.quartic.weyl.core.model.*
import rx.Observable
import java.time.Clock

class SpatialPredicateComputation @JvmOverloads constructor(
        private val layerId: LayerId,
        private val spatialPredicateSpec: SpatialPredicateSpec,
        private val clock: Clock = Clock.systemUTC()
) : LayerPopulator {

    override fun dependencies() = listOf(spatialPredicateSpec.layerA, spatialPredicateSpec.layerB)

    override fun spec(dependencies: List<Layer>): LayerSpec {
        val layerA = dependencies[0]
        val layerB = dependencies[1]

        return LayerSpec(
                layerId,
                LayerMetadata(
                        name(layerA, layerB, spatialPredicateSpec),
                        name(layerA, layerB, spatialPredicateSpec),
                        layerA.spec.metadata.attribution,
                        clock.instant()
                ),
                IDENTITY_VIEW,
                layerA.spec.staticSchema,
                true
        )
    }

    private fun name(layerA: Layer, layerB: Layer, spatialPredicateSpec: SpatialPredicateSpec)
            = "${layerA.spec.metadata.name} ${spatialPredicateSpec.predicate} ${layerB.spec.metadata.name}"

    override fun updates(dependencies: List<Layer>): Observable<LayerUpdate> {
        val (_, _, _, _, indexedFeatures) = dependencies[0]
        val (_, features) = dependencies[1]


        val bufferedFeatures = indexedFeatures
                .filter { (preparedGeometry) ->
                    features.stream()
                            .anyMatch { (_, geometry) -> spatialPredicateSpec.predicate.test(preparedGeometry, geometry) }
                }
                .map { feature -> NakedFeature(
                        feature.feature.entityId.uid,
                        feature.feature.geometry,
                        feature.feature.attributes
                )}

        return Observable.never<LayerUpdate>().startWith(LayerUpdate(REPLACE, bufferedFeatures))
    }
}
